packages/react-core/src/components/Menu/MenuItem.tsx

```diff
@@ -77,20 +80,80 @@ const MenuItemBase: React.FunctionComponent<MenuItemProps> = ({
   icon,
   actions,
   onShowFlyout,
-  innerRef,
   drilldownMenu,
   isOnPath,
   ...props
 }: MenuItemProps) => {
+  const {
+    menuId,
+    parentMenu,
+    onSelect,
+    onActionClick,
+    activeItemId,
+    selected,
+    drilldownItemPath,
+    onDrillIn,
+    onDrillOut
+  } = React.useContext(MenuContext);
   const Component = component || to ? 'a' : 'button';
   const [flyoutVisible, setFlyoutVisible] = React.useState(false);
   const [flyoutTarget, setFlyoutTarget] = React.useState(null);
+  const flyoutContext = React.useContext(FlyoutContext);
+  const [flyoutXDirection, setFlyoutXDirection] = React.useState(flyoutContext.direction);
+  const [offsetX, setOffsetX] = React.useState(0);
+  const [offsetY, setOffsetY] = React.useState(0);
+  const ref = React.useRef<HTMLLIElement>();
 
+  const hasFlyout = flyoutMenu !== undefined;
   const showFlyout = (displayFlyout: boolean) => {
     setFlyoutVisible(displayFlyout);
     onShowFlyout && displayFlyout && onShowFlyout();
   };
 
+  React.useLayoutEffect(() => {
+    if (hasFlyout && ref.current && canUseDOM) {
+      const flyoutMenu = ref.current.lastElementChild;
+      if (flyoutMenu && flyoutMenu.classList.contains(styles.menu)) {
+        const origin = ref.current.getClientRects()[0];
+        const rect = flyoutMenu.getClientRects()[0];
+        if (origin && rect) {
+          const spaceLeftLeft = origin.x - rect.width;
+          const spaceLeftRight = window.innerWidth - origin.x - origin.width - rect.width;
+          if (spaceLeftLeft < 0 && spaceLeftRight < 0) {
+            // Calculate x offsets
+            setOffsetX(-spaceLeftLeft);
+          } else if (spaceLeftRight < 0) {
+            setFlyoutXDirection('left');
+          } else if (spaceLeftLeft < 0) {
+            setFlyoutXDirection('right');
+          }
+          const spaceLeftBot = window.innerHeight - origin.y - rect.height;
+          // This is calculated with respect to spaceLeftBot
+          const spaceLeftTop = window.innerHeight - rect.height;
+          if (spaceLeftTop < 0 && spaceLeftBot < 0) {
+            // Idea: can scroll down, but not up
+            setOffsetY(0);
+          } else if (spaceLeftBot < 0) {
+            setOffsetY(-spaceLeftBot);
+          }
+        }
+      }
+    }
+  }, [flyoutVisible, flyoutMenu]);
+
+  React.useEffect(() => {
+    const flyoutMenu = ref.current.lastElementChild as HTMLElement;
+    if (flyoutMenu && flyoutMenu.classList.contains(styles.menu)) {
+      flyoutMenu.style.right = flyoutXDirection === 'right' ? 'auto' : `calc(100% - ${offsetX}px`;
+      flyoutMenu.style.left = flyoutXDirection === 'left' ? 'auto' : `calc(100% - ${offsetX}px`;
+      flyoutMenu.style.top = `calc(-${offsetY}px + var(--pf-c-menu--m-flyout__menu--Top))`;
```

Comment:
Can you try doing this via 

```suggestion
      flyoutMenu.style.setProperty = `--pf-c-menu--m-flyout__menu--top-offset: ${offsetY}px`;
```

I don't know if this syntax is right, but basically passing the offset to the var we created and allowing core to handle the `calc()` - https://github.com/patternfly/patternfly/blob/6fe562844cf6e99990603c2885de1911a37fd9b7/src/patternfly/components/Menu/menu.scss#L10

Reply:
Done.